[MT-1749] Add user identifier & custom variables support, cleanup, and dependency updates

sebastian-krajna opened the pull request:
- Added support for tracking user IDs
- Added support for tracking custom variables
- Updates Gradle and SDK dependencies

sebastian-krajna committed:
Fix documentation link in README.md for Contentsquare integration

sebastian-krajna committed:
Update dependencies and configurations for Android SDK and Gradle wrapper

- Updated Kotlin, Tealium, and AndroidX library versions in build.gradle files.
- Increased compileSdk and targetSdk versions to 33.
- Added new entries to .gitignore for local settings and IDE configurations.
- Updated Gradle wrapper to version 7.6.4.

sebastian-krajna committed:
Update Gradle configurations and dependencies

- Increased JVM memory allocation in gradle.properties.
- Added AndroidProjectSystem and runConfigurations.xml files for IDE support.
- Updated AndroidX test dependencies in app/build.gradle.
- Adjusted test implementation version for Robolectric in contentsquare/build.gradle.
- Enabled debugging for the contentsquare module.

sebastian-krajna committed:
Add user identifier and custom variables tracking to Contentsquare integration

- Introduced methods for sending user identifiers and custom variables in ContentsquareCommand and ContentsquareInstance.
- Updated MainActivity to include options for sending user identifiers and custom variables.
- Enhanced contentsquare.json with new keys for user identifier and custom variables.
- Added corresponding tests for user identifier and custom variables functionality in ContentsquareRemoteCommandTest.
- Updated AndroidManifest.xml and strings.xml for new features and configurations.

sebastian-krajna committed:
Remove unused .idea configuration files and update .gitignore to include .idea directory. This cleanup enhances project structure by eliminating unnecessary IDE-specific files.

sebastian-krajna committed:
Refine custom variable validation in ContentsquareInstance

- Updated the validation logic for custom variables to ensure the index is within the range of 1 to 20.

jameskeith commented on contentsquare/src/main/java/com/tealium/remotecommands/contentsquare/ContentsquareInstance.kt:
```diff
@@ -42,6 +44,33 @@ class ContentsquareInstance(private val application: Application) : Contentsquar
+        Contentsquare.sendUserIdentifier(userId)
+    }
+
+    override fun sendCustomVars(screenName: String, customVars: Array<JSONObject>) {
+        Log.d(TAG, "Sending custom variables for screen: $screenName")
+        val csCustomVars = customVars.mapNotNull { json ->
+            val index = json.optInt(CustomVars.INDEX, -1)
+            val name = json.optString(CustomVars.NAME)
+            val value = json.optString(CustomVars.VALUE)
+
+            if (index in 1..20) {
+                CustomVar(index, name, value)
+            } else {
+                Log.e(TAG, "Invalid custom var index: $index (must be 1..20). Skipping: $json")
+                null
+            }
+        }.toTypedArray()
+        
+        if (csCustomVars.isNotEmpty()) {
+            Contentsquare.send(screenName, csCustomVars)
+        } else {
+            Log.d(TAG, "No valid custom vars to send")
+        }
+    }
```
think this was already raised in the iOS PR, but this command should probably just be combined with the regular `send` command for screenviews, since they are related.
Make the `customVars` array optional, and use that to decide which method to call.

jameskeith commented on contentsquare/src/main/java/com/tealium/remotecommands/contentsquare/ContentsquareInstance.kt:
```diff
@@ -42,6 +44,33 @@ class ContentsquareInstance(private val application: Application) : Contentsquar
+            val name = json.optString(CustomVars.NAME)
+            val value = json.optString(CustomVars.VALUE)
+
+            if (index in 1..20) {
```
saw that iOS was recommending not being so restrictive on this too, which seems fair.
perhaps when converting this, we can simply fail early when we get the index:
```kotlin
 val index = json.optInt(CustomVars.INDEX, -1)
 if (index <= 0) return@mapNotNull null
 // ...
```

jameskeith commented on contentsquare/build.gradle:
```diff
@@ -3,11 +3,12 @@ apply plugin: 'kotlin-android'
-        minSdkVersion 19
-        targetSdkVersion 32
+        minSdkVersion 21
+        targetSdkVersion 33
```
similar to the other RemoteCommand prs, but this should probably be targeting 35 for longevity.

sebastian-krajna replied:
Of course not! it's just me.. :) 
maybe someday I will be good friend with gradle, but not yet haha

jameskeith commented on contentsquare/build.gradle:
```diff
@@ -3,11 +3,12 @@ apply plugin: 'kotlin-android'
     defaultConfig {
-        minSdkVersion 19
-        targetSdkVersion 32
+        minSdkVersion 21
```
was there a new minimum requirement that meant we needed to update this?

sebastian-krajna replied:
Manifest merger failed : uses-sdk:minSdkVersion 19 cannot be smaller than version 21 declared in library [com.tealium:kotlin-remotecommand-dispatcher:1.5.2]

jameskeith replied:
but this module doesn't depend on the `remotecommand-dispatcher` - so as long as this specific module doesn't require 21, then it can safely stay as 19.
I would expect that error came from leaving the `app` module as 19 (which does depend on the remotecommand-dispatcher.

jameskeith commented on app/src/main/java/com/tealium/example/MainActivity.kt:
```diff
@@ -69,13 +72,42 @@ class MainActivity : AppCompatActivity() {
+        
+        private fun sendCustomVars() {
+            val customVar1 = JSONObject().apply {
+                put("index", 0)
```
this is indexed as `0` but the current loop inside our implementation expects this to be `1 .. 20` - can we confirm what the expected range of values is for ContentSquare and align this example with the expected value.

jameskeith reviewed (commented):
looks like a good start so far

sebastian-krajna committed:
Update Gradle configurations and enhance Contentsquare integration

- Upgraded Gradle and Android plugin versions to 8.6.0 and 8.7 respectively.
- Increased compileSdk and targetSdk versions to 35 in app and contentsquare modules.
- Refactored ContentsquareCommand and ContentsquareInstance to support sending custom variables with screen views.
- Removed deprecated sendCustomVars method and updated related tests for new functionality.

sebastian-krajna committed:
Refactor custom variable handling in Contentsquare integration

- Updated the JSON structure for custom variables to an object of arrays format, aligning with iOS implementation.
- Modified MainActivity to utilize the new custom variable format when sending events.
- Adjusted ContentsquareConstants to reflect new keys for custom variable indexes, names, and values.
- Enhanced ContentsquareRemoteCommand to parse the new custom variable format and updated related tests accordingly.

jameskeith commented on contentsquare/src/main/java/com/tealium/remotecommands/contentsquare/ContentsquareConstants.kt:
```diff
@@ -9,6 +9,8 @@ object Commands {
     const val SEND_TRANSACTION = "sendtransaction"
     const val SEND_DYNAMIC_VAR = "senddynamicvar"
+    const val SEND_USER_IDENTIFIER = "senduseridentifier"
+    const val SEND_CUSTOM_VARS = "sendcustomvars"
```
this is no longer used, as it's part of the `sendscreenview` command.

jameskeith commented on build.gradle:
```diff
@@ -1,20 +1,20 @@
+    ext.tealium_remotecommand_dispatcher_version = '1.5.2'
+    ext.tealium_tagmanagement_dispatcher_version = '1.2.5'
+    ext.tealium_lifecycle_version = '1.2.2'
     ext.tealium_contentsquare_version = '2.2.1'
```
this is going to need to be increased. probably to 2.3.0 as there dont appear to be any breaking changes.

jameskeith commented on build.gradle:
```diff
@@ -1,20 +1,20 @@
+    ext.tealium_lifecycle_version = '1.2.2'
     ext.tealium_contentsquare_version = '2.2.1'
-    ext.contentsquare_version = ":[4.9, 5.0)"
+    ext.contentsquare_version = ":[4.38, 5.0)"
```
was this jump necessary?

sebastian-krajna replied:
Shouln't we update this to latest? 

https://docs.contentsquare.com/en/android/changelog/#4382---20250516

jameskeith commented on contentsquare/src/main/java/com/tealium/remotecommands/contentsquare/ContentsquareInstance.kt:
```diff
@@ -12,8 +14,31 @@ class ContentsquareInstance(private val application: Application) : Contentsquar
-    override fun send(screenName: String) {
-        Contentsquare.send(screenName)
+    override fun send(screenName: String, customVars: Array<JSONObject>?) {
+        if (customVars != null && customVars.isNotEmpty()) {
+            Log.d(TAG, "Sending custom variables for screen: $screenName")
+            val csCustomVars = customVars.mapNotNull { json ->
+                val index = json.optInt("index", -1)
+                val name = json.optString("name")
+                val value = json.optString("value")
+
+                if (index > 0) {
+                    CustomVar(index, name, value)
+                } else {
+                    Log.e(TAG, "Invalid custom var index: $index (must be positive). Skipping: $json")
+                    null
+                }
+            }.toTypedArray()
+            
+            if (csCustomVars.isNotEmpty()) {
+                Contentsquare.send(screenName, csCustomVars)
+            } else {
+                Log.d(TAG, "No valid custom vars to send, sending regular screen view")
+                Contentsquare.send(screenName)
+            }
+        } else {
+            Contentsquare.send(screenName)
+        }
```
if it was me, I'd probably try and tidy up some of this nesting. Something like:

```kotlin
override fun send(screenName: String, customVars: Array<JSONObject>?) {
    if (!customVars.isNullOrEmpty()) {
        Log.d(TAG, "Sending custom variables for screen: $screenName")
        val csCustomVars = customVars.mapNotNull(::jsonObjectToCustomVar)
            .toTypedArray()
        
        if (csCustomVars.isNotEmpty()) {
            Contentsquare.send(screenName, csCustomVars)
            return
        }
    }

    Contentsquare.send(screenName)
}

private fun jsonObjectToCustomVar(json: JSONObject): CustomVar? {
    val index = json.optInt("index", -1)
    if (index < 0) {
        Log.e(TAG, "Invalid custom var index: $index (must be positive). Skipping: $json")
        return null
    }

    val name = json.optString("name")
    val value = json.optString("value")

    return CustomVar(index, name, value)
}
```

jameskeith commented on contentsquare/src/test/java/com/tealium/remotecommands/contentsquare/ContentsquareRemoteCommandTest.kt:
```diff
@@ -186,4 +258,104 @@ class ContentsquareRemoteCommandTest {
+        confirmVerified(mockCommand)
+    }
+    
+    @Test
```
great work for spotting and adding all these missing tests.

jameskeith reviewed (approved):
changes look good so far; nice work
i've noted a couple of other questions, but happy for it to be released once they've been addressed.

sebastian-krajna committed:
Update Contentsquare integration and Gradle configurations

- Bumped tealium_contentsquare_version to 2.3.0 for improved functionality.
- Lowered minSdkVersion to 19 in contentsquare module for broader compatibility.
- Refactored custom variable handling in ContentsquareInstance to streamline the sending process and improve validation logic.

sebastian-krajna committed:
Update Contentsquare version range in build.gradle
